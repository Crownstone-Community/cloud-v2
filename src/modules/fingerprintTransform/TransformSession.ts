import {Util} from "../../util/Util";
import {TransformUtil} from "./TransformUtil";
import {User} from "../../models/user.model";
import {Sphere} from "../../models/sphere.model";
import {Dbs} from "../containers/RepoContainer";
import {EventHandler} from "../sse/EventHandler";


export class TransformSession {
  id: uuid;
  sphereId: string;
  destructor : () => void;
  cleanupTimeout : NodeJS.Timeout;

  userA: User;
  userB: User;
  sphere: Sphere

  userId_A: string;
  userId_B: string;
  deviceId_A: string;
  deviceId_B: string;

  invitePending: boolean = false;

  datasets: {[uuid:string] : {setA: MeasurementMap, setB: MeasurementMap}} = {};

  inviteInterval : NodeJS.Timeout;
  inviteTimeout  : NodeJS.Timeout;

  constructor(id: uuid, sphereId: string, destructor: () => void, userId_A: string, deviceId_A: string, userId_B: string, deviceId_B: string) {
    this.id = id;
    this.sphereId = sphereId;
    this.destructor = destructor;

    this.cleanupTimeout = setTimeout(() => {
      this.cleanup();
    }, 30*60*1000); // 30 minutes

    this.userId_A = userId_A;
    this.userId_B = userId_B;
    this.deviceId_A = deviceId_A;
    this.deviceId_B = deviceId_B;
  }

  async init() : Promise<void> {
    try {
      this.userA = await Dbs.user.findById(this.userId_A);
      this.userB = await Dbs.user.findById(this.userId_B);
      this.sphere = await Dbs.sphere.findById(this.sphereId);

      this.invitePending = true;
      this.inviteInterval = setInterval(() => {
        if (this.invitePending === false) {
          this._stopInviting();
        }

        // This acts as an invitation. The other user will join the session by calling the joinSession function.
        EventHandler.transform.sendTransformSessionRequestedEvent(
          this.sphere,
          this.id,
          this.userA,
          this.userB,
          this.deviceId_A,
          this.deviceId_B
        );
      }, 2000);

      this.inviteTimeout = setTimeout(() => {
        this._stopInviting();
      }, 60 * 1000); // 1 minute
    }
    catch (err: any) {
      console.log("Failed to init transform session", err);
      this.cleanup();
      throw err;
    }

  }

  _stopInviting() {
     clearTimeout(this.inviteTimeout);
     clearInterval(this.inviteInterval);
     if (this.invitePending === true) {
       this.cleanup();
     }
  }

  joinSession(userId: string) {
    if (this.userId_B === userId) {
      this.invitePending = false;
      this._stopInviting();
      EventHandler.transform.sendTransformSessionReadyEvent(
        this.sphere,
        this.id,
        this.userA,
        this.userB,
        this.deviceId_A,
        this.deviceId_B
      );
    }
    else {
      throw new Error("User is not invited to this session.");
    }
  }

  cleanup() {
    clearTimeout(this.inviteTimeout);
    clearInterval(this.inviteInterval);
    clearTimeout(this.cleanupTimeout);
    this.destructor();

    EventHandler.transform.sendTransformSessionStoppedEvent(
      this.sphere,
      this.id,
    );
  }

  startDatasetCollection() : uuid {
    let datasetId = Util.getUUID();
    this.datasets[datasetId] = {setA: null, setB: null};
    EventHandler.transform.sendTransformDatacollectionStartedEvent(
      this.sphere,
      this.id,
      datasetId,
    );
    return datasetId;
  }

  finishedCollectingDataset(userId: string, datasetUUID: uuid, deviceId: string, dataset: MeasurementMap) {
    if (this.datasets[datasetUUID] === undefined) { throw new Error("Dataset not found"); }

    if (userId === this.userId_A && this.deviceId_A == this.deviceId_A && this.datasets[datasetUUID].setA === null) {
      this.datasets[datasetUUID].setA = dataset;
      if (this.datasets[datasetUUID].setB === null) {
        EventHandler.transform.sendTransformDatacollectionReceivedEvent(
          this.sphere,
          this.id,
          datasetUUID,
          this.userA,
          this.deviceId_A
        );
      }
    }
    else if (userId === this.userId_A && this.deviceId_A == this.deviceId_A && this.datasets[datasetUUID].setA !== null) {
      throw new Error("Already received dataset from user A for this collection session.");
    }
    else if (userId === this.userId_B && this.deviceId_B == this.deviceId_B && this.datasets[datasetUUID].setB === null) {
      this.datasets[datasetUUID].setB = dataset;
      if (this.datasets[datasetUUID].setA === null) {
        EventHandler.transform.sendTransformDatacollectionReceivedEvent(
          this.sphere,
          this.id,
          datasetUUID,
          this.userB,
          this.deviceId_B
        );
      }
    }
    else if (userId === this.userId_B && this.deviceId_B == this.deviceId_B && this.datasets[datasetUUID].setB !== null) {
      throw new Error("Already received dataset from user B for this collection session.");
    }

    if (this.datasets[datasetUUID].setA !== null && this.datasets[datasetUUID].setB !== null) {
      let qualityResult = getDatasetQuality(this.datasets);
      EventHandler.transform.sendTransformDatacollectionFinishedEvent(
        this.sphere,
        this.id,
        datasetUUID,
        qualityResult
      );
    }
  }

  generateTransformSets() : TransformResult {
    let aToB = this._generateTransformSetDirection(true);
    let bToA = this._generateTransformSetDirection(false);
    let result = [
      {sessionId: this.id, fromUser: this.userA.id, fromDevice: this.deviceId_A, toUser: this.userB.id, toDevice: this.deviceId_B, transform: aToB},
      {sessionId: this.id, fromUser: this.userB.id, fromDevice: this.deviceId_B, toUser: this.userA.id, toDevice: this.deviceId_A, transform: bToA}
    ];
    EventHandler.transform.sendTransformSetFinishedEvent(
      this.sphere,
      this.id,
      result,
    );
    return result;
  }


  _generateTransformSetDirection(aToB:boolean) {
    let fromMaps : MeasurementMap[] = [];
    let toMaps   : MeasurementMap[] = [];
    for (let datasetId in this.datasets) {
      if (this.datasets[datasetId].setA !== null && this.datasets[datasetId].setB !== null) {
        if (aToB) {
          fromMaps.push(this.datasets[datasetId].setA);
          toMaps.push(this.datasets[datasetId].setB);
        }
        else {
          fromMaps.push(this.datasets[datasetId].setB);
          toMaps.push(this.datasets[datasetId].setA);
        }
      }
    }

    if (fromMaps.length === 0) {
      throw new Error("No complete datasets found.");
    }

    return this._generateTransformSet(fromMaps, toMaps);
  }

  _generateTransformSet(sets_From:MeasurementMap[], sets_To:MeasurementMap[]) : TransformSet {
    return TransformUtil.getTransFormSet(sets_From, sets_To);
  }
}

function getDatasetQuality(datasets: {[uuid:string] : {setA: MeasurementMap, setB: MeasurementMap}}) : {userA: Record<string, number>, userB: Record<string, number>} {
  let comparisonArray_AtoB : TransformArray = [];
  let comparisonArray_BtoA : TransformArray = [];
  for (let uuid in datasets) {
    if (datasets[uuid].setA === null || datasets[uuid].setB === null) { continue; }

    let rawMap_AtoB = TransformUtil.getRawMap_AtoB(datasets[uuid].setA, datasets[uuid].setB)
    let rawMap_BtoA = TransformUtil.getRawMap_AtoB(datasets[uuid].setB, datasets[uuid].setA)
    comparisonArray_AtoB = comparisonArray_AtoB.concat(rawMap_AtoB);
    comparisonArray_BtoA = comparisonArray_BtoA.concat(rawMap_BtoA);
  }

  comparisonArray_AtoB.sort((a,b) => { return b[0] - a[0]; });
  comparisonArray_BtoA.sort((a,b) => { return b[0] - a[0]; });
  let coreBuckets       = [ -50, -55, -60, -65, -70, -75, -80, -85, -90 ];

  let coreA = checkBucketFillFactor(coreBuckets, comparisonArray_AtoB);
  let coreB = checkBucketFillFactor(coreBuckets, comparisonArray_BtoA);

  return {userA: coreA, userB: coreB};
}

function checkBucketFillFactor(buckets:number[], data: TransformArray) : Record<string,number> {
  let bucketedData = TransformUtil.fillBuckets(buckets, data);

  // check if there are at least 4 buckets with at least 3 datapoints.
  let quality : Record<string, number> = {}
  for (let data of bucketedData) {
    quality[data.x] = data.data.length;
  }

  for (let bucket of buckets) {
    if (!quality[bucket]) {
      quality[bucket] = 0;
    }
  }
  return quality;
}
