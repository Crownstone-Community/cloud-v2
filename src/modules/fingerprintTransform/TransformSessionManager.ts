import {TransformSession} from "./TransformSession";
import {Util} from "../../util/Util";


class TransformSessionManagerClass {
  activeSessions : Record<uuid, TransformSession> = {};


  /**
   * The sphereId is used for authenticated routing of sse events
   * @param sphereId
   * @param userId_A
   * @param phoneType_A
   * @param userId_B
   * @param phoneType_B
   */
  async createNewSession(sphereId: string, userId_A: string, phoneType_A: string, userId_B: string, phoneType_B: string) : Promise<uuid> {
    let id = Util.getUUID();
    let session = new TransformSession(
      id,
      sphereId,
      () => { delete this.activeSessions[id]; },
      userId_A, phoneType_A, userId_B, phoneType_B
    );

    this.activeSessions[session.id] = session;

    await session.init(); // this can throw an error.
    return id;
  }

  killSession(sessionUUID : uuid) {
    if (this.activeSessions[sessionUUID]) {
      this.activeSessions[sessionUUID].cleanup();
    }
  }

  joinSession(sessionUUID : uuid, userId: string) {
    if (this.activeSessions[sessionUUID]) {
      this.activeSessions[sessionUUID].joinSession(userId);
    }
    else {
      throw new Error("Session not found");
    }
  }

  startDatasetCollection(sessionUUID: uuid) : uuid {
    if (this.activeSessions[sessionUUID]) {
      return this.activeSessions[sessionUUID].startDatasetCollection();
    }
    throw new Error("Session not found");
  }

  finishedCollectingDataset(sessionUUID : uuid, datasetUUID: uuid, userId: string, dataset: MeasurementMap) {
    if (this.activeSessions[sessionUUID]) {
      this.activeSessions[sessionUUID].finishedCollectingDataset(userId, datasetUUID, dataset);
    }
    else {
      throw new Error("Session not found");
    }
  }

  generateTransformSets(sessionUUID : uuid) : TransformResult {
    if (this.activeSessions[sessionUUID]) {
      return this.activeSessions[sessionUUID].generateTransformSets();
    }
    throw new Error("Session not found");
  }

}


export const TransformSessionManager = new TransformSessionManagerClass();