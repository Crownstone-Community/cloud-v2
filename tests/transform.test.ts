import {TransformUtil} from "../src/modules/fingerprintTransform/TransformUtil";

test("Test dataparsing", async () => {
  let phoneA_data : MeasurementMap = {
    1: -71,
    2: -62,
    3: -56,
    4: -75,
    5: -85,
    6: -94,
    7: -72,
    8: -72,
    9: -40,
  }

  let phoneB_data : MeasurementMap = {
    1: -78,
    2: -65,
    3: -58,
    4: -79,
    5: -84,
    6: -90,
    7: -72,
    8: -74,
    9: -43,
  }

  let projection    : TransformArray = TransformUtil.getRawMap_AtoB(phoneA_data, phoneB_data);
  let normalizedMap : TransformArray = TransformUtil.getNormalizedMap(projection);
  let buckets = TransformUtil.getBuckets();
  let bucketedData = TransformUtil.fillBuckets(buckets, normalizedMap);
  let bucketedAverages = TransformUtil.getAveragedBucketMap(bucketedData);
  let interpolatedValues = TransformUtil.getInterpolatedValues(bucketedAverages);

  let transformSet = [...bucketedAverages, ...interpolatedValues].filter((item) => { return item.data[0] !== null });
  transformSet.sort((a,b) => { return b.x - a.x });

  // console.log('projection', projection);
  // console.log('normalizedMap', normalizedMap);
  // console.log('bucketedData', bucketedData);
  // console.log('bucketedAverages',bucketedAverages)
  // console.log('interpolatedValues',interpolatedValues)
  // console.log('transformSet',transformSet)

  expect(transformSet).toMatchSnapshot();

  let transformedSet = TransformUtil.transformDataset(Object.values(phoneA_data), transformSet);
  // console.log('phoneA_data',Object.values(phoneA_data))
  // console.log('phoneB_data',Object.values(phoneB_data))
  // console.log('transformedSet',transformedSet)
  let error = Object.values(phoneB_data).map((value, index) => { return value - transformedSet[index] });
  console.log(error)
  expect(transformedSet).toMatchSnapshot();
});

test("With actual data", async () => {
  let data_iphone_1 = {
    '12399_Min:26559': -70.11111111111111,
    '15433_Min:6247': -75.55555555555556,
    '17452_Min:54653': -87,
    '18652_Min:50338': -74,
    '20161_Min:10414': -74.8,
    '20675_Min:31553': -79,
    '23546_Min:6110': -81.44444444444444,
    '24959_Min:41807': -87.5,
    '26910_Min:44873': -53.9,
    '302_Min:11314': -60,
    '3145_Min:34773': -70.875,
    '32562_Min:44536': -59.9,
    '34060_Min:7621': -77.88888888888889,
    '36655_Min:39097': -69.6,
    '38643_Min:9749': -87.2,
    '46798_Min:59413': -80.22222222222223,
    '47254_Min:57646': -79,
    '47912_Min:57777': -90,
    '51826_Min:3597': -73.5,
    '53424_Min:12784': -83.44444444444444,
    '56053_Min:25176': -70.4,
    '64810_Min:33239': -86.66666666666667,
    '8569_Min:45914': -73.625,
    '9524_Min:56756': -79
  }
  let data_ipad_1 = {
    '12399_Min:26559': -67.77777777777777,
    '15433_Min:6247': -72.33333333333333,
    '17452_Min:54653': -79.375,
    '18652_Min:50338': -69.2,
    '20161_Min:10414': -69.5,
    '20675_Min:31553': -75.6,
    '23546_Min:6110': -79,
    '24959_Min:41807': -79,
    '26910_Min:44873': -52.4,
    '302_Min:11314': -57.2,
    '3145_Min:34773': -69,
    '32562_Min:44536': -66.8,
    '34060_Min:7621': -77,
    '36655_Min:39097': -73.11111111111111,
    '38643_Min:9749': -77.66666666666667,
    '46798_Min:59413': -78.44444444444444,
    '47254_Min:57646': -72,
    '47912_Min:57777': -81.5,
    '51826_Min:3597': -75.71428571428571,
    '53424_Min:12784': -83,
    '56053_Min:25176': -68.4,
    '64810_Min:33239': -79.4,
    '8569_Min:45914': -72.42857142857143,
    '9524_Min:56756': -74.125
  }

  let data_ipad_2 = {
    '11282_Min:27384': -85.14285714285714,
    '12399_Min:26559': -57.1,
    '15433_Min:6247': -75.55555555555556,
    '15968_Min:23173': -90.33333333333333,
    '17452_Min:54653': -82.25,
    '18652_Min:50338': -73.71428571428571,
    '19567_Min:61824': -89.2,
    '20161_Min:10414': -60.8,
    '20675_Min:31553': -76.77777777777777,
    '23546_Min:6110': -78.77777777777777,
    '24511_Min:2529': -86.5,
    '24959_Min:41807': -81.11111111111111,
    '26910_Min:44873': -45.4,
    '302_Min:11314': -56,
    '3145_Min:34773': -67.8,
    '32562_Min:44536': -70.3,
    '34060_Min:7621': -77.5,
    '34819_Min:14002': -87.66666666666667,
    '36655_Min:39097': -61.3,
    '38643_Min:9749': -77.125,
    '46798_Min:59413': -86.71428571428571,
    '47254_Min:57646': -79.625,
    '47912_Min:57777': -84.8,
    '50001_Min:43540': -90,
    '51826_Min:3597': -72.1,
    '53424_Min:12784': -77.1,
    '56053_Min:25176': -63.6,
    '64810_Min:33239': -76.5,
    '8569_Min:45914': -63.1,
    '9524_Min:56756': -86.5
  }
  let data_iphone_2 ={
    '12399_Min:26559': -72.6,
    '15433_Min:6247': -69.4,
    '17452_Min:54653': -87.22222222222223,
    '18652_Min:50338': -82.11111111111111,
    '20161_Min:10414': -70,
    '20675_Min:31553': -79.9,
    '23546_Min:6110': -73.9,
    '24959_Min:41807': -82.75,
    '26910_Min:44873': -46.9,
    '302_Min:11314': -61.7,
    '3145_Min:34773': -72.3,
    '32562_Min:44536': -76.4,
    '34060_Min:7621': -78.7,
    '36655_Min:39097': -65.3,
    '38643_Min:9749': -88.14285714285714,
    '46798_Min:59413': -87.11111111111111,
    '47254_Min:57646': -78.2,
    '47912_Min:57777': -87.4,
    '51826_Min:3597': -75.5,
    '53424_Min:12784': -74.4,
    '56053_Min:25176': -62.1,
    '64810_Min:33239': -81.88888888888889,
    '8569_Min:45914': -68.7,
    '9524_Min:56756': -86.75
  }

  let sets_From = [data_ipad_1, data_ipad_2];
  let sets_To = [data_iphone_1, data_iphone_2];

  let ipadToIphone = TransformUtil.getTransFormSet(sets_From, sets_To);
  let iphoneToIpad = TransformUtil.getTransFormSet(sets_To, sets_From);

  expect(ipadToIphone).toMatchSnapshot();
  expect(iphoneToIpad).toMatchSnapshot();
});
