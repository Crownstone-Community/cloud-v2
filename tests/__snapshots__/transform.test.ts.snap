// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test dataparsing 1`] = `
Array [
  Object {
    "data": Array [
      -10,
      1.148660714285714,
    ],
    "x": -10,
  },
  Object {
    "data": Array [
      -17,
      1.131473214285714,
    ],
    "x": -17,
  },
  Object {
    "data": Array [
      -24,
      1.114285714285714,
    ],
    "x": -24,
  },
  Object {
    "data": Array [
      -31,
      1.0970982142857142,
    ],
    "x": -31,
  },
  Object {
    "data": Array [
      -40,
      1.075,
    ],
    "x": -38,
  },
  Object {
    "data": Array [
      -45,
      1.0627232142857144,
    ],
    "x": -45,
  },
  Object {
    "data": Array [
      -56,
      1.0357142857142858,
    ],
    "x": -52,
  },
  Object {
    "data": Array [
      -62,
      1.0483870967741935,
    ],
    "x": -59,
  },
  Object {
    "data": Array [
      -72,
      1,
    ],
    "x": -66,
  },
  Object {
    "data": Array [
      -75,
      1.0533333333333332,
    ],
    "x": -73,
  },
  Object {
    "data": Array [
      -85,
      0.9882352941176471,
    ],
    "x": -80,
  },
  Object {
    "data": Array [
      -87,
      0.9813934084272007,
    ],
    "x": -87,
  },
  Object {
    "data": Array [
      -94,
      0.9574468085106383,
    ],
    "x": -94,
  },
]
`;

exports[`Test dataparsing 2`] = `
Array [
  -71.34354838709677,
  -65,
  -58.00000000000001,
  -78.99999999999999,
  -84,
  -90,
  -72,
  -72,
  -43,
]
`;

exports[`With actual data 1`] = `
Array [
  Object {
    "data": Array [
      -10,
      0.899250613196589,
    ],
    "x": -10,
  },
  Object {
    "data": Array [
      -17,
      0.9257060719723947,
    ],
    "x": -17,
  },
  Object {
    "data": Array [
      -24,
      0.9521615307482005,
    ],
    "x": -24,
  },
  Object {
    "data": Array [
      -31,
      0.9786169895240063,
    ],
    "x": -31,
  },
  Object {
    "data": Array [
      -38,
      1.0050724482998121,
    ],
    "x": -38,
  },
  Object {
    "data": Array [
      -45.4,
      1.0330396475770924,
    ],
    "x": -45,
  },
  Object {
    "data": Array [
      -56.6,
      1.0753683816183817,
    ],
    "x": -52,
  },
  Object {
    "data": Array [
      -62.2,
      1.0770004369149153,
    ],
    "x": -59,
  },
  Object {
    "data": Array [
      -70.07142857142857,
      1.0618838234684627,
    ],
    "x": -66,
  },
  Object {
    "data": Array [
      -77.54149659863945,
      1.0414138761094436,
    ],
    "x": -73,
  },
  Object {
    "data": Array [
      -84.1375,
      1.024839463231463,
    ],
    "x": -80,
  },
  Object {
    "data": Array [
      -87,
      1.017646583850377,
    ],
    "x": -87,
  },
  Object {
    "data": Array [
      -94,
      1.000057009818027,
    ],
    "x": -94,
  },
]
`;

exports[`With actual data 2`] = `
Array [
  Object {
    "data": Array [
      -10,
      0.9461214995787034,
    ],
    "x": -10,
  },
  Object {
    "data": Array [
      -17,
      0.9502751284658078,
    ],
    "x": -17,
  },
  Object {
    "data": Array [
      -24,
      0.9544287573529121,
    ],
    "x": -24,
  },
  Object {
    "data": Array [
      -31,
      0.9585823862400165,
    ],
    "x": -31,
  },
  Object {
    "data": Array [
      -38,
      0.9627360151271208,
    ],
    "x": -38,
  },
  Object {
    "data": Array [
      -46.9,
      0.9680170575692963,
    ],
    "x": -45,
  },
  Object {
    "data": Array [
      -53.9,
      0.9721706864564007,
    ],
    "x": -52,
  },
  Object {
    "data": Array [
      -61.23333333333333,
      1.015654693356084,
    ],
    "x": -59,
  },
  Object {
    "data": Array [
      -70.54768518518519,
      0.9614388564830367,
    ],
    "x": -66,
  },
  Object {
    "data": Array [
      -76.94444444444446,
      0.963915761040731,
    ],
    "x": -73,
  },
  Object {
    "data": Array [
      -83.37222222222223,
      0.9445906479998282,
    ],
    "x": -80,
  },
  Object {
    "data": Array [
      -87.4752380952381,
      0.937309419218305,
    ],
    "x": -87,
  },
  Object {
    "data": Array [
      -94,
      0.9257305500050553,
    ],
    "x": -94,
  },
]
`;
